### PR TITLE
Build Class for Credential Exposure

### PR DESCRIPTION
**Pull request by DJensen94**
## 🗣 Description ##

Builds out a Class for credential exposure data, making it simpler to access variables from the pages script without passing so many variables

## 💭 Motivation and context ##
This makes the report generator more modular, allowing for data to be accessed anywhere more simply without passing a ton of variables to every script

## 🧪 Testing ##

This current PR needs to pass current pytest /tests

## ✅ Checklist ##

- [ ] This PR has an informative and human-readable title.
- [ ] Changes are limited to a single goal - _eschew scope creep!_
- [ ] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
- [ ] All relevant type-of-change labels have been added.
- [ ] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [ ] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [ ] All new and existing tests pass.

### CODE REVIEW

**Comment by cduhn17** [comment]
@DJensen94 
When you get a chance take a look at the LGTM notice. Also be sure to maintain an imperative mood on [commit](https://chris.beams.io/posts/git-commit/#imperative) messages.

**Review by aloftus23 — APPROVED** [review]

File: src/pe_reports/metrics.py [hunk 1]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
         self.start_date = start_date
         self.end_date = end_date
         self.org_uid = org_uid
-        self.query_cyberSix_creds = query_cyberSix_creds(org_uid, start_date, end_date)
+        c6_df = query_cyberSix_creds(org_uid, start_date, end_date)
+        c6_df.loc[
+            c6_df["breach_name"] == "", "breach_name"
+        ] = "Cyber_six_" + pd.to_datetime(c6_df["breach_date"]).dt.strftime("%m/%d/%Y")
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: This line is pretty gnarly and might benefit from being broken up into more than one line of code.

File: src/pe_reports/metrics.py [hunk 2]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+        c6_df["description"] = (
+            c6_df["description"].str.split("Query to find the related").str[0]
+        )
+        c6_df["password_included"] = np.where(c6_df["password"] != "", True, False)
+        self.query_cyberSix_creds = c6_df
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Does it make any sense to move the cyberSix code above into `query_cyberSix_creds()` or is it specific to this metrics code?

File: src/pe_reports/metrics.py [hunk 3]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
 
-    # TODO the following functions correspond to functions at report_generator.py
-    # TODO and will be added in follow up PR found here: https://github.com/cisagov/pe-reports/pull/99
-    def password(self):
-        """Return total number of credentials with passwords."""
 
-    def breached(self):
-        """Return total number of breached credentials."""
+class Domains_Masqs:
+    """Domains Masquerading Class."""
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Unnecessary capitalization:
```suggestion
    """Domains Masquerading class."""
```

File: src/pe_reports/metrics.py [hunk 4]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
-    def breach_details(self):
-        """Return breach details."""
+    def count(self):
+        """Return total count of malicious domains."""
+        df = self.df_mal
+        return len(df.index)
+
+    def summary(self):
+        """Return domain masquerading summary information."""
+        df_mal = self.df_mal
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: This line seems unnecessary to me- you can just use `self.df_mal` throughout this function.

File: src/pe_reports/metrics.py [hunk 5]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+        return risky_assets
+
+    def risky_ports_count(self):
+        """Return total count of insecure protocols."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: You've got a lot of repeated code in these three functions, namely:
```
insecure_df = self.insecure_df
insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
insecure = insecure[
    (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
]
risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```
You should consider moving that code to a separate function.

File: src/pe_reports/metrics.py [hunk 6]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+            verif_vulns["count"] = verif_vulns["ip"].str.split("  ").str.len()
+            verifVulns = verif_vulns["count"].sum()
+
+        else:
+            verifVulns = 0
+
+        return verifVulns
+
+    def unverified_cve(self):
+        """Return top 15 unverified cves and their counts."""
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Use proper capitalization for acronyms in docstrings:
```suggestion
        """Return top 15 unverified CVEs and their counts."""
```

File: src/pe_reports/metrics.py [hunk 7]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+            .reset_index(drop=True)
+        )
+        unverif_df["count"] = unverif_df["potential_vulns"].str.split(",").str.len()
+        unverif_df = unverif_df[["ip", "count"]]
+        unverif_df = unverif_df.sort_values(by=["count"], ascending=False)
+        unverif_df = unverif_df[:15].reset_index(drop=True)
+        return unverif_df
+
+    def unverified_vuln_count(self):
+        """Return the count of ips with unverified vulnerabilities."""
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Use proper capitalization for acronyms in docstrings:
```suggestion
        """Return the count of IP addresses with unverified vulnerabilities."""
```

File: src/pe_reports/metrics.py [hunk 8]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+        unverif_df = (
+            unverif_df[["potential_vulns", "ip"]]
+            .drop_duplicates(keep="first")
+            .reset_index(drop=True)
+        )
+
+        return len(unverif_df.index)
+
+    def verif_vulns(self):
+        """Return a dataframe with each CVE, the associated ips and the affected ports."""
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Use proper capitalization for acronyms in docstrings:
```suggestion
        """Return a dataframe with each CVE, the associated IPs and the affected ports."""
```

File: src/pe_reports/metrics.py [hunk 9]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+                "ip": "IP",
+                "port": "Port",
+                "summary": "Summary",
+            }
+        )
+        return verif_vulns_summary
+
+
+class Cyber_Six:
+    """Dark Web and Cyber Six Data Class."""
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Remove unnecessary capitalization:
```suggestion
    """Dark web and Cyber Six data class."""
```

File: src/pe_reports/metrics.py [hunk 10]
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
...
+        )
+        dark_web_tags["tags"] = dark_web_tags["tags"].str.strip("{}")
+        dark_web_tags["tags"] = dark_web_tags["tags"].str.replace(
+            ",", ", ", regex=False
+        )
+        dark_web_tags = dark_web_tags.rename(columns={"tags": "Tags"})
+        return dark_web_tags
+
+    def top_cve_table(self):
+        """Get top cves."""
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Use proper capitalization for acronyms in docstrings:
```suggestion
        """Get top CVEs."""
```

**Review by dav3r — CHANGES_REQUESTED** [review]
Please take a look at my questions, suggestions, and various minor nitpicks - thanks!

File: src/pe_reports/metrics.py (hunk 2 above)
Comment (DJensen94) [reply src/pe_reports/metrics.py]: @dav3r comment has been addressed in the following commit: [d8560b5](https://github.com/cisagov/pe-reports/pull/99/commits/d8560b50ae42d0e6d21fea02bb78526ad7076b4c)
Moved code to db_query.py as suggested

File: src/pe_reports/metrics.py (hunk 1 above)
Comment (DJensen94) [reply src/pe_reports/metrics.py]: @dav3r comment has been addressed in the following commit: [d8560b5](https://github.com/cisagov/pe-reports/pull/99/commits/d8560b50ae42d0e6d21fea02bb78526ad7076b4c)

File: src/pe_reports/metrics.py (hunk 5 above)
Comment (DJensen94) [reply src/pe_reports/metrics.py]: @dav3r Addressed this comment in the following commit: [051e84b]( https://github.com/cisagov/pe-reports/pull/99/commits/051e84b856e0c6bf33b31d24fcdaf704ac199406)

File: src/pe_reports/metrics.py (hunk 2 above)
Comment (dav3r) [reply src/pe_reports/metrics.py]: @DJensen94 - Just a reminder that our convention is to allow the person who starts a conversation be the one to resolve it (the exception is when you accept commit suggestions and GitHub automatically resolves the conversation).

File: src/pe_reports/metrics.py (hunk 1 above)
Comment (dav3r) [reply src/pe_reports/metrics.py]: That definitely looks better!

File: src/pe_reports/metrics.py (hunk 4 above)
Comment (DJensen94) [reply src/pe_reports/metrics.py]: @dav3r, I've addressed this comment in the following commit: [cc99f8b](https://github.com/cisagov/pe-reports/pull/99/commits/cc99f8b7733cb76a25af69a6b49eee2b7cfd9bc9)

File: src/pe_reports/metrics.py [hunk 11]
```diff
@@ -15,22 +24,552 @@ def __init__(self, start_date, end_date, org_uid):
...
+
+    @staticmethod
+    def isolate_risky_assets(df):
+        """Return risky assets from the insecure_df dataframe."""
+        insecure = df[df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
+        return risky_assets
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Remove unnecessary variable:
```suggestion
        return insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```

File: src/pe_reports/metrics.py [hunk 12]
```diff
@@ -15,22 +24,552 @@ def __init__(self, start_date, end_date, org_uid):
...
+        # Horizontal bar: insecure protocol count
+        pro_count = risky_assets.groupby(["protocol"], as_index=False)["protocol"].agg(
+            {"id_count": "count"}
+        )
+        return pro_count
+
+    def risky_assets(self):
+        """Return dataframe of ip and risky protocols."""
+        risky_assets = self.isolate_risky_assets(self.insecure_df)
+        return risky_assets
```
Comment (dav3r) [inline src/pe_reports/metrics.py]: Is this function really necessary if all it does is call `self.isolate_risky_assets()` and return its output?

**Review by dav3r — CHANGES_REQUESTED** [review]
Thanks for addressing my previous round of feedback- I just have a couple more items based on your recent changes.

File: src/pe_reports/metrics.py (hunk 12 above)
Comment (DJensen94) [reply src/pe_reports/metrics.py]: @dav3r, we can probably remove this and just call the static method (isolate_risky_assets) instead, as long as we are ok to call the static method while using the class

Comment (dav3r) [reply src/pe_reports/metrics.py]: I'm fine with that approach.

Comment (DJensen94) [reply src/pe_reports/metrics.py]: @dav3r addressed suggestion in the following commit: [6858068](https://github.com/cisagov/pe-reports/pull/99/commits/68580680776f164eb2e8b144afc603029093c0a5)

Comment (dav3r) [reply src/pe_reports/metrics.py]: Thanks!

**Review by dav3r — APPROVED** [review]
Looking good here! 👍 👍

**Review by cduhn17 — APPROVED** [review]
Looks good.

**Pull request merged**